site/app/controllers/admin/ReportController.php

```diff
@@ -776,8 +772,56 @@ public function displayGradebook() {
             new WebResponse(
                 ['admin', 'Report'],
                 'showFullGradebook',
-                $grade_file
+                $grade_file,
+                $grade_summaries_last_run
             )
         );
     }
+
+    /**
+     * Generate a custom filename for the downloaded CSV file
+     *
+     * @return string
+     */
+    private function generateCustomFilename() {
+        $course = $this->core->getConfig()->getCourse();
+        $timestamp = date("ymdHis");
+        return "{$course}_rainbow_grades_{$timestamp}.csv";
+    }
+
+
+    /**
+     * Download CSV file for Rainbow Grades
+     * @return void
+     */
+    #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_csv")]
+    public function downloadRainbowGradesCSVFile() {
+        // Check if the user has the necessary permissions (e.g., admin access)
+        if (!$this->core->getUser()->accessAdmin()) {
+            $this->core->getOutput()->showError("This account cannot access admin pages");
+        }
+
+        // Path to the CSV file for Rainbow Grades
+        $csvFilePath = '/var/local/submitty/courses/' . $this->core->getConfig()->getTerm() . '/' . $this->core->getConfig()->getCourse() . '/rainbow_grades/output.csv';
+
+        // Check if the file exists
+        if (file_exists($csvFilePath)) {
+            // Set appropriate headers for file download with a custom filename
+            header('Content-Description: File Transfer');
+            header('Content-Type: application/csv');
+            header('Content-Disposition: attachment; filename="' . $this->generateCustomFilename() . '"');
+            header('Expires: 0');
+            header('Cache-Control: must-revalidate');
+            header('Pragma: public');
+            header('Content-Length: ' . filesize($csvFilePath));
+
+            // Output the file content
+            readfile($csvFilePath);
+            exit;
+        }
+        else {
+            // Handle the case where the file does not exist
+            $this->core->getOutput()->showError($csvFilePath . " was not found or was not readable.\nMaybe you have not <a\thref='https://submitty.org/instructor/course_settings/rainbow_grades/automatic_setup'>generated the rainbow grades</a> yet?");
```

Comment:
Putting a link in an error message banner seems like a bad idea...

Reply:
Why?

Reply:
If the banner message disappears after a period of time, so too will the documentation the users wants to click on.  Couldn't you display this else in the UI somewhere instead?  At a minimum, the HTML should be injected via Twig.

Reply:
The error message will never disappear, as it is being rendered by ErrorPage.twig, which never hides the error message. I believe that the only way to move the link out of file would be to either to change how error pages are created in Submitty or make a brand-new error page specifically for this one error. I could also just remove the link, but I figured having it there would make it easier for users to debug their problem.